src/Persistence/AbstractQuery.php

```diff
@@ -0,0 +1,389 @@
+    {
+        $this->scope->addCondition(...func_get_args());
+
+        return $this;
+    }
+
+    public function whereId($id)
+    {
+        if (!$this->model->id_field) {
+            throw (new Exception('Unable to find record by "id" when Model::id_field is not defined.'))
+                ->addMoreInfo('id', $id);
+        }
+
+        $this->where($this->model->getField($this->model->id_field), $id);
+
+        return $this;
+    }
+
+    abstract protected function initWhere(): void;
+
+    public function order($field, $desc = null): self
+    {
+        $this->order[] = [$field, $desc];
+
+        $this->initOrder();
+
+        return $this;
+    }
+
+    abstract protected function initOrder(): void;
+
+    public function limit($limit, $offset = 0): self
+    {
+        $this->limit = [$limit, $offset];
+
+        $this->initLimit();
+
+        return $this;
+    }
+
+    abstract protected function initLimit(): void;
+
+    /**
+     * Converts limit array to arguments [$limit, $offset].
+     */
+    protected function getLimitArgs()
+    {
+        if ($this->limit) {
+            $offset = $this->limit[1] ?? 0;
+            $limit = $this->limit[0] ?? null;
+
+            if ($limit || $offset) {
+                if ($limit === null) {
+                    $limit = PHP_INT_MAX;
+                }
+
+                return [$limit, $offset ?? 0];
+            }
+        }
+    }
+
+    public function execute()
+    {
+        return $this->executeQueryWithDebug(function () {
+            $this->withMode();
+
+            // backward compatibility
+            $this->hookOnModel('HOOK_BEFORE_' . $this->getMode(), [$this]);
+
+//          $this->model->hook(self::HOOK_BEFORE_EXECUTE, [$this]);
+
+            $this->hookOnModel('HOOK_BEFORE_' . $this->getMode(), [$this]);
+
+            $result = $this->doExecute();
+
+            // backward compatibility
+            $this->hookOnModel('HOOK_AFTER_' . $this->getMode(), [$this, $result]);
+
+//          $this->model->hook(self::HOOK_AFTER_EXECUTE, [$this, $result]);
+
+            return $result;
+        });
+    }
+
+    abstract protected function doExecute();
+
+    protected function hookOnModel($name, $args = []): void
+    {
+        $hookSpotConst = self::class . '::' . $name;
+        if (defined($hookSpotConst)) {
+            // backward compatibility
+//             if ($this->model->hookHasCallbacks(constant($hookSpotConst))) {
+//                 'trigger_error'('Hook spot deprecated. Use AbstractQuery::HOOK_BEFORE_EXECUTE or AbstractQuery::HOOK_AFTER_EXECUTE instead', E_USER_DEPRECATED);
+//             }
+            $this->model->hook(constant($hookSpotConst), $args);
+        }
+    }
+
+    /**
+     * Get array of records matching the query.
+     */
+    public function get(): array
+    {
+        return $this->executeQueryWithDebug(function () {
+            return $this->doGet();
+        });
+    }
+
+    abstract protected function doGet(): array;
```

Comment:
I would prefer `_` prefix instead even if not advised by PSR - it is then much clearer what it does.

Reply:
I don't like the `_` prefix as then we have several methods with same name `get`, `_get`, `__get`.
Not very descriptive.
The `do` prefix is more clear and is already used in UserActions for instance.

Reply:
not critical to me...

Reply:
`doGet` naming looks fine.